File: src/android/DialogWifi.java
package com.siemens.plugins.DialogWifi;

import org.apache.cordova.CordovaPlugin;
import org.apache.cordova.CallbackContext;
import org.apache.cordova.PluginResult;
import org.apache.cordova.PluginResult.Status;
import android.util.Log;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.ArrayList;
import java.util.List;

/**
 * This class echoes a string called from JavaScript.
 */
public class DialogWifi extends CordovaPlugin {

  private static final String TAG = DialogWifi.class.getSimpleName();
  private final DialogWifiImpl mDialogWifiImpl = new DialogWifiImpl();
  private CallbackContext mConnectCallback;
  private CallbackContext mCloseCallback;
  private CallbackContext mWifiListCallback;
  private CallbackContext mDPMConfigCallback;
  private CallbackContext mWifiConfigCallback;

  private String randomNo = "";
  private String serialNo = "";

  private final DialogWifiImpl.ISocketStatusCallback mCallback = new DialogWifiImpl.ISocketStatusCallback() {
    @Override
    public void onConnectResult(boolean connected) {
      if (mConnectCallback != null) {
        if (connected) {
          sendConnected();
          PluginResult result = new PluginResult(Status.OK, "");
          result.setKeepCallback(true);
          mConnectCallback.sendPluginResult(result);
        } else {
          mConnectCallback.error("");
        }
      }
    }

    @Override
    public void onClosed() {
      if (mCloseCallback != null) {
        mCloseCallback.success();
      }
      if (mConnectCallback != null) {
        mConnectCallback.error("");
      }
    }

    @Override
    public void onDateReport(JSONObject jsonObject) {
      Log.i(TAG, "== onDateReport ==" + jsonObject.toString());
      if (jsonObject.has("SOCKET_TYPE")) {
        receiveSocketType(jsonObject);
      }

      if (jsonObject.has("thingName")) {
        receiveThingName(jsonObject);
      }

      if (jsonObject.has("mode")) {
        receiveMode(jsonObject);
      }

      if (jsonObject.has("APList")) {
        receiveAPList(jsonObject);
      }

      if (jsonObject.has("SET_AP_SSID_PW")) {
        receiveSetApSSIDPW(jsonObject);
      }

      if (jsonObject.has("randomNo")) {
        receiveSerialNo(jsonObject);
      }

      if (jsonObject.has("RESULT_REBOOT")) {
        receiveRebootResult(jsonObject);
      }
    }
  };

  @Override
  public boolean execute(String action, JSONArray args, CallbackContext callbackContext) throws JSONException {
    switch (action) {
      case "onConnectSocket":
        String host = args.getString(0);
        int port = args.getInt(1);
        this.onConnectSocket(host, port, callbackContext);
        return true;
      case "sendDPMSet":
        this.sendDPMSet(callbackContext);
        return true;
      case "close":
        this.onClose(callbackContext);
        return true;
      case "scan":
        this.scan(callbackContext);
        return true;
      case "sendSSIDPW":
        String _ssid = args.getString(0);
        String _pwd = args.getString(1);
        int _security = args.getInt(2);
        int _isHidden = args.getInt(3);
        String _url = args.getString(4);
        this.sendSSIDPW(_ssid, _pwd, _security, _isHidden, _url, callbackContext);
        return true;
    }
    return false;
  }

  /**
   * Socket Connect to AP Server after Device Wifi AP Connection Established.
   */
  public void onConnectSocket(String host, int port, CallbackContext callbackContext) {
    Log.i(TAG, "onConnectSocket()" + host + " " + port);
    mConnectCallback = callbackContext;
    mDialogWifiImpl.setHost(host, port);
    mDialogWifiImpl.onInitSocket(mCallback);
  }

  public void onClose(CallbackContext callbackContext) {
    Log.i(TAG, "onClose()");
    mCloseCallback = callbackContext;
    mDialogWifiImpl.onClose();
  }

  public void scan(CallbackContext callbackContext) {
    Log.i(TAG, "scan()");
    mWifiListCallback = callbackContext;
    final JSONObject obj = new JSONObject();
    try {
      obj.put("msgType", 3);
      obj.put("REQ_RESCAN", 0);
    } catch (Exception e) {
      e.printStackTrace();
    }
    mDialogWifiImpl.write(obj, ex -> {
      if (ex != null) {
        Log.e(TAG, "Sending message error");
        ex.printStackTrace();
      } else {
        Log.i(TAG, "Sending message Completed");
      }
    });
  }

  public void sendConnected() {
    final JSONObject obj = new JSONObject();
    try {
      obj.put("msgType", 0);
      obj.put("CONNECTED", 0);
    } catch (Exception e) {
      e.printStackTrace();
    }
    mDialogWifiImpl.write(obj, ex -> {
      if (ex != null) {
        Log.e(TAG, "Sending message error");
        ex.printStackTrace();
      } else {
        Log.i(TAG, "Sending message Completed");
      }
    });
  }

  public void sendDPMSet(CallbackContext callbackContext) {
    Log.i(TAG, "sendDPMSet ");
    mDPMConfigCallback = callbackContext;
    final JSONObject obj = new JSONObject();
    try {
      obj.put("msgType", 5);
      obj.put("REQ_SET_DPM", 0);
      obj.put("sleepMode", 0);
      obj.put("rtcTimer", 1740);
      obj.put("useDPM", 0);
      obj.put("dpmKeepAlive", 30000);
      obj.put("userWakeUp", 0);
      obj.put("timWakeup", 10);
    } catch (Exception e) {
      e.printStackTrace();
    }

    mDialogWifiImpl.write(obj, ex -> {
      if (ex != null) {
        Log.e(TAG, "Sending message error");
        ex.printStackTrace();
      } else {
        Log.i(TAG, "Sending message Completed");
      }
    });
  }

  public void sendSSIDPW(String _ssid, String _pwd, int _security, int _isHidden, String _url,
      CallbackContext callbackContext) {
    Log.i(TAG, "sendSSIDPW");
    this.mWifiConfigCallback = callbackContext;
    final JSONObject obj = new JSONObject();
    try {
      obj.put("msgType", 1);
      obj.put("SET_AP_SSID_PW", 0);
      obj.put("ssid", _ssid);
      obj.put("pw", _pwd);
      obj.put("securityType", _security);
      obj.put("isHidden", _isHidden);
      obj.put("url", _url);
    } catch (Exception e) {
      e.printStackTrace();
    }
    mDialogWifiImpl.write(obj, ex -> {
      if (ex != null) {
        Log.e(TAG, "Sending message error");
        ex.printStackTrace();
      } else {
        Log.i(TAG, "Sending message Completed");
      }
    });
  }

  public void receiveSetApSSIDPW(JSONObject obj) {
    if (obj != null) {
      Log.i(TAG, "== receiveSetApSSIDPW ==");
      try {
        if (obj.getInt("SET_AP_SSID_PW") != -1) {
          Log.i(TAG, "SET_AP_SSID_PW = " + obj.getInt("SET_AP_SSID_PW"));
          return;
        }
      } catch (Exception e) {
        e.printStackTrace();
      }
    }
    if (mWifiConfigCallback != null) {
      mWifiConfigCallback.error("SET_AP_SSID_PW error");
    }
  }

  public void receiveSocketType(JSONObject obj) {
    if (obj != null) {
      Log.i(TAG, "== receiveSocketType() ==");
      try {
        Log.i(TAG, "socketType = " + obj.getInt("SOCKET_TYPE"));

      } catch (Exception e) {
        e.printStackTrace();
      }

    }
  }

  public void receiveThingName(JSONObject obj) {
    if (obj != null) {
      Log.i(TAG, "== receiveThingName() ==");
      try {
        Log.i(TAG, "thingName = " + obj.getString("thingName"));
      } catch (Exception e) {
        e.printStackTrace();
      }

    }
  }

  public void receiveMode(JSONObject obj) {
    if (obj != null) {
      Log.i(TAG, "== receiveMode() ==");
      try {
        Log.i(TAG, "mode = " + obj.getInt("mode"));
      } catch (Exception e) {
        e.printStackTrace();
      }

    }
  }

  public void receiveRebootResult(JSONObject obj) {

    if (obj != null) {
      Log.i(TAG, "== receiveRebootResult ==");
      try {
        if (obj.getInt("RESULT_REBOOT") != -1) {
          Log.i(TAG, "RESULT_REBOOT = " + obj.getInt("RESULT_REBOOT"));
          mDialogWifiImpl.onClose();
          if (mWifiConfigCallback != null) {
            List<PluginResult> resultList = new ArrayList<PluginResult>();
            PluginResult randomNoResult = new PluginResult(PluginResult.Status.OK, randomNo);
            PluginResult serialNoResult = new PluginResult(PluginResult.Status.OK, serialNo);
            resultList.add(randomNoResult);
            resultList.add(serialNoResult);
            mWifiConfigCallback.sendPluginResult(new PluginResult(PluginResult.Status.OK, resultList));
            return;
          }
        }
      } catch (Exception e) {
        e.printStackTrace();
      }
    }
    if (mWifiConfigCallback != null) {
      mWifiConfigCallback.error("reboot error");
    }
  }

  public void receiveAPList(JSONObject jsonObject) {

    Log.i(TAG, "== receiveAPList() ==");
    JSONArray jsonArray;
    try {
      jsonArray = jsonObject.getJSONArray("APList");
      if (mWifiListCallback != null) {
        mWifiListCallback.success(jsonArray);
      }
    } catch (JSONException e) {
      e.printStackTrace();
      if (mWifiListCallback != null) {
        mWifiListCallback.error(e.getMessage());
      }
    }
  }

  public void receiveSerialNo(JSONObject obj) {
    if (obj != null) {
      Log.i(TAG, "== receiveSerialNo() ==");
      try {
        randomNo = obj.getString("randomNo");
        serialNo = obj.getString("SN");
        Log.i(TAG, "randomNo = " + randomNo);
        Log.i(TAG, "serialNo = " + serialNo);
      } catch (Exception e) {
        e.printStackTrace();
      }

    }
  }
}

File: src/ios/DialogWifi.swift
import Foundation
import NetworkExtension
import SystemConfiguration.CaptiveNetwork

@objc(DialogWifi) public class DialogWifi : CDVPlugin, GCDAsyncSocketDelegate  {

    var wifiConfiguration: NEHotspotConfiguration?

    var mSocket: GCDAsyncSocket!
    var mTimer: Timer?
    var timeCount: Int!
    var isCompleted = false
    var randomNo = ""
    var serialNo = ""
    
    var currentNetworkInfos: Array<NetworkInfo>? {
        get {
            return SSID.fetchNetworkInfo()
        }
    }
    
    override public func pluginInitialize() {
        timeCount = 0
    }

    override public func onAppTerminate() {
    }

    
    var connectCommand: CDVInvokedUrlCommand!
    @objc public func onConnectSocket(_ command: CDVInvokedUrlCommand) {

        self.connectCommand = command
        let host = command.argument(at: 0) as! String
        let port = command.argument(at: 1) as! UInt16
        let ssid = command.argument(at: 2) as! String
        let pw = command.argument(at: 3) as! String

        self.connectWifi(ssidS: ssid, pw: pw, host: host, port: port)
    }
    
    var disconnectCommand: CDVInvokedUrlCommand!
    @objc public func close(_ command: CDVInvokedUrlCommand) {

        self.disconnectCommand = command
        self.disconnectSocket()
    }
    
    var scanCommand: CDVInvokedUrlCommand!
    @objc public func scan(_ command: CDVInvokedUrlCommand) {
        self.scanCommand = command
        self.tcpSendReqRescan()
    }
    
    @objc public func sendDPMSet(_ command: CDVInvokedUrlCommand) {
        self.tcpSendDPMSet()
        let pluginResult = CDVPluginResult(status:CDVCommandStatus_OK, messageAs: "OK")
        self.commandDelegate?.send(pluginResult, callbackId: command.callbackId)
    }
    
    var configCommand: CDVInvokedUrlCommand!
    @objc public func sendSSIDPW(_ command: CDVInvokedUrlCommand) {
        self.configCommand = command
        let ssid = command.argument(at: 0) as! String
        let pwd = command.argument(at: 1) as! String
        let security = command.argument(at: 2) as! Int
        let hidden = command.argument(at: 3) as! Int
        let url = command.argument(at: 4) as! String
        let bind = command.argument(at: 5) as! Int
        self.tcpSendSSIDPW(ssid: ssid, pw: pwd, security: security, isHidden: hidden, serverURL: url, bind: bind)
    }
    
    // MARK: - WIFI Connection
    
    public func connectWifi(ssidS: String, pw: String, host: String, port: UInt16) {
        if (pw.isEmpty) {
            wifiConfiguration = NEHotspotConfiguration(ssid: ssidS)
        } else {
            wifiConfiguration = NEHotspotConfiguration(
                ssid: ssidS,
                passphrase: pw,
                isWEP: false)
        }
        wifiConfiguration?.joinOnce = true;
        NEHotspotConfigurationManager.shared.apply(wifiConfiguration!) { error in
            if let ssid = self.currentNetworkInfos?.first?.ssid {
                print("connected SSID: \(ssid)")
                if ssid == ssidS {
                    DispatchQueue.main.asyncAfter(deadline: DispatchTime.now() + 5, execute: {
                        self.connectSocket(host:host, port:port)
                    })
                } else {
                    if (self.connectCommand != nil) {
                        let pluginResult = CDVPluginResult(status:CDVCommandStatus_ERROR, messageAs: "error")
                        self.commandDelegate?.send(pluginResult, callbackId: self.connectCommand.callbackId)
                        self.connectCommand = nil
                    }

                }
            } else {
                if (self.connectCommand != nil) {
                    let pluginResult = CDVPluginResult(status:CDVCommandStatus_ERROR, messageAs: "error")
                    self.commandDelegate?.send(pluginResult, callbackId: self.connectCommand.callbackId)
                    self.connectCommand = nil
                }
            }
            
        }
    }
    
    public struct NetworkInfo {
        var interface: String
        var success: Bool = false
        var ssid: String?
        var bssid: String?
    }
    
    public class SSID {
        class func fetchNetworkInfo() -> [NetworkInfo]? {
            if let interfaces: NSArray = CNCopySupportedInterfaces() {
                var networkInfos = [NetworkInfo]()
                for interface in interfaces {
                    let interfaceName = interface as! String
                    var networkInfo = NetworkInfo(interface: interfaceName,
                                                  success: false,
                                                  ssid: nil,
                                                  bssid: nil)
                    if let dict = CNCopyCurrentNetworkInfo(interfaceName as CFString) as NSDictionary? {
                        networkInfo.success = true
                        networkInfo.ssid = dict[kCNNetworkInfoKeySSID as String] as? String
                        networkInfo.bssid = dict[kCNNetworkInfoKeyBSSID as String] as? String
                    }
                    networkInfos.append(networkInfo)
                }
                return networkInfos
            }
            return nil
        }
    }
    
    // MARK: - TCP Connection
    
    public func connectSocket(host: String, port: UInt16) {
        do {
            mSocket = GCDAsyncSocket(delegate: self, delegateQueue: DispatchQueue.main)
            try mSocket.connect(toHost: host, onPort: port, withTimeout: 20)
        } catch let error {
            print(error)
            if self.connectCommand != nil {
                let pluginResult = CDVPluginResult(status:CDVCommandStatus_ERROR, messageAs: "error")
                self.commandDelegate?.send(pluginResult, callbackId: self.connectCommand.callbackId)
            }
        }
    }
    
    public func disconnectSocket() {
        print("==> disconnectSocket()\n")
        if mSocket != nil {
            print("==>self.mSocket != nil\n")
            if mSocket.isConnected {
                print("==>self.mSocket.isConnected \n")
                mSocket.disconnect()
            }
            mSocket = nil
        }
    }
    
    public func socket(_ sock: GCDAsyncSocket, didAcceptNewSocket newSocket: GCDAsyncSocket) {
        print(">> didAcceptNewSocket!\n")
        mSocket = newSocket
    }
    
    public func socket(_ socket: GCDAsyncSocket, didConnectToHost host: String, port p:UInt16) {
        print(">> didConnectToHost!\n")
        let sslSettings = [
            GCDAsyncSocketManuallyEvaluateTrust: NSNumber(value: true),
            kCFStreamSSLPeerName: "www.apple.com"
        ] as? [CFString : Any]
        mSocket.startTLS(sslSettings as? [String:NSObject])
    }
    
    public func socket(_ sock: GCDAsyncSocket, didWriteDataWithTag tag: Int) {
        print(">> didWriteData");
    }
    
    public func socket(_ sock: GCDAsyncSocket, didReceive trust: SecTrust, completionHandler: @escaping (Bool) -> Void) {
        print(">> didReceiveData")
        completionHandler(true)
    }
    
    public func socket(_ sock: GCDAsyncSocket, didRead: Data, withTag tag:CLong){
        print(">> didRead!");
        do {
            let json = try JSON(data: didRead)
            print("incoming message: \(json))")
            
            if json["SOCKET_TYPE"].exists() {
                let socketType: String? = json["SOCKET_TYPE"].stringValue
                   print("socketType : \(String(describing: socketType!))")
                   if socketType == "0" || socketType == "1" {
                       mSocket.disconnect()
                   }
            }

            if json["thingName"].exists() {
                let thingName = json["thingName"].stringValue
                print("thingname : \(thingName)")
                // UserDefaults.standard.set(thingName, forKey: "thingNameKey")
            }

            if json["APList"].exists() {
                if self.scanCommand != nil {
                    print("APList : \(json["APList"].arrayValue)")
                    var apList = [NSDictionary]();
                    for item in json["APList"].arrayValue {
                        let index = item["index"].intValue
                        let SSID = item["SSID"].stringValue
                        let securityType = item["securityType"].intValue
                        let signal = item["signal"].intValue
                        let ap: NSDictionary = NSDictionary(
                           objects: [index, SSID, securityType, signal],
                           forKeys: ["index" as NSCopying, "SSID" as NSCopying, "securityType" as NSCopying, "signal" as NSCopying])
                        apList.append(ap);
                    }
                    let pluginResult = CDVPluginResult(status:CDVCommandStatus_OK, messageAs: apList)
                    self.commandDelegate?.send(pluginResult, callbackId: self.scanCommand.callbackId)
                    self.scanCommand = nil
                }
            }
            
            if json["randomNo"].exists() {
                randomNo = json["randomNo"].stringValue
                serialNo = json["SN"].stringValue
                print("randomNo : \(randomNo)")
                print("serialNo : \(serialNo)")
            }
            
            if json["RESULT_REBOOT"].exists() {
                let resultReboot:String? = json["RESULT_REBOOT"].stringValue
                print("resultReboot = \(String(describing: resultReboot!))")
                if (self.configCommand == nil) {
                    return
                }
                if resultReboot == "0"{
                    var array = [String]();
                    array.append(randomNo)
                    array.append(serialNo)
                    let pluginResult = CDVPluginResult(status:CDVCommandStatus_OK, messageAsMultipart: array)
                    self.commandDelegate?.send(pluginResult, callbackId: self.configCommand.callbackId)
                } else {
                    let pluginResult = CDVPluginResult(status:CDVCommandStatus_ERROR, messageAs: "error")
                    self.commandDelegate?.send(pluginResult, callbackId: self.configCommand.callbackId)
                }
            }
            
        } catch let error {
            print(error)
        }
        
    }
    
    public func socketDidSecure(_ sock: GCDAsyncSocket) {
        print(">> socketDidSecure!")
        
        tcpSendConnected()
        
    }
    
    
    public func socketDidDisconnect(_ sock: GCDAsyncSocket, withError err: Error?) {
        print(">> didDisconnect!")
        if (!isCompleted) {
//                showToast(message: "Socket is not connected. Please retry.", seconds: 2.0)
        }
        if let error = err {
            print("Will disconnect with error: \(error)")
        } else{
            print("Success disconnect")
        }
        if mSocket != nil {
            print("==>mSocket != nil\n")
            if mSocket.isConnected {
                print("==>mSocket.isConnected \n")
                mSocket.disconnect()
            }
            mSocket = nil
        }
        if self.disconnectCommand != nil {
            let pluginResult = CDVPluginResult(status:CDVCommandStatus_OK, messageAs: "didDisconnect")
            self.commandDelegate?.send(pluginResult, callbackId: self.disconnectCommand.callbackId)
            self.disconnectCommand = nil;
        }
        if self.connectCommand != nil {
            let pluginResult = CDVPluginResult(status:CDVCommandStatus_ERROR, messageAs: "didDisconnect")
            self.commandDelegate?.send(pluginResult, callbackId: self.connectCommand.callbackId)
            self.connectCommand = nil;
        }
    }
    
    
    //MARK: - Send TCP
    
//    func tcpSendSocketType(type: Int) {
//        print("==> tcpSendSocketType()\n")
//        let cmdSocketType = "{\"msgType\":6,\"SOCKET_TYPE\":\(type)}"
//        let data = cmdSocketType.data(using: String.Encoding.utf8)!
//        mSocket.write(data, withTimeout:10, tag: 0)
//        mSocket.readData(withTimeout: -1, tag: 0)
//
//    }
    
    func tcpSendConnected() {
        print("==> tcpSendConnected()\n")
        if (mSocket == nil) {
            return
        }
        let cmdConnected: String = "{\"msgType\":0,\"CONNECTED\":0}"
        let data = cmdConnected.data(using: String.Encoding.utf8)!
        mSocket.write(data, withTimeout:10, tag: 0)
        mSocket.readData(withTimeout: -1, tag: 0)
        if (self.connectCommand != nil) {
            let pluginResult = CDVPluginResult(status:CDVCommandStatus_OK, messageAs: "OK")
            pluginResult?.setKeepCallbackAs(true)
            self.commandDelegate?.send(pluginResult, callbackId: self.connectCommand.callbackId)
        }
    }
    
    func tcpSendReqRescan() {
        print("==> tcpSendReqRescan()\n")
        if (mSocket == nil) {
            return
        }
        let cmdReqRescan: String = "{\"msgType\":3,\"REQ_RESCAN\":0}"
        let data = cmdReqRescan.data(using: String.Encoding.utf8)!
        mSocket.write(data, withTimeout:10, tag: 0)
        mSocket.readData(withTimeout: -1, tag: 0)

    }
    
    func tcpSendDPMSet() {
        let cmdSetDpm: String = "{\"msgType\":5, \"REQ_SET_DPM\":0, \"sleepMode\":0, \"rtcTimer\":1740, \"useDPM\":0, \"dpmKeepAlive\":30000, \"userWakeUp\":0, \"timWakeup\":10}"
        print("tcpSendDPMSet : \(cmdSetDpm)")
        if (mSocket == nil) {
            return
        }
        let data = cmdSetDpm.data(using: String.Encoding.utf8)!
        mSocket.write(data, withTimeout:10, tag: 0)
        mSocket.readData(withTimeout: -1, tag: 0)
        
    }
    
    func tcpSendSSIDPW(ssid: String, pw: String, security: Int, isHidden: Int, serverURL: String, bind: Int) {
        let cmdSendSSIDPW: String = "{\"msgType\":1, \"SET_AP_SSID_PW\":0, \"ssid\":\"\(ssid)\", \"pw\":\"\(pw)\", \"securityType\":\(security),  \"isHidden\":\(isHidden),  \"bind\":\(bind), \"url\":\"\(serverURL)\"}"
        print("tcpSendSSIDPW : \(cmdSendSSIDPW)")
        if (mSocket == nil) {
            return
        }
        let data = cmdSendSSIDPW.data(using: String.Encoding.utf8)!
        mSocket.write(data, withTimeout:10, tag: 0)
        mSocket.readData(withTimeout: -1, tag: 0)
    }
    
}
